### PR TITLE
Stream responses from node

### PR DESCRIPTION
**Pull request by KirillPamPam**
Now we can stream responses from a node using stream parser with the following algorithm:

- Our aim is to get a result or an error object/field and pass it further - to accomplish it we count brackets or quotes to determine the start and the end of this object/field.
- First of all we parse the first part of bytes to determine if we have an error or a result. If we have a result we determine whether it's a scalar value or an object/array and start to count quotes/brackets. If we have an error we just parse it.
- If we couldn't parse the first part we return `AggregateResponse` where we aggregate all bytes just as we did before.
- Then we see if we parsed the whole response or not. If we did so we return it otherwise we start streaming the rest bytes counting brackets/quotes.
- When we determine the last chunk we point it out setting `finalChunk` to true. Otherwise if we couldn't determine the end of response for some reason we set `finalChunk` deliberately to end the stream.

Now we turn on streaming only for `eth_call` and `eth_getLogs`. Also we can control this process via `dproxy` and field `chunkSize` - if it is not specified we turn off streaming.

### CODE REVIEW

File: src/main/kotlin/io/emeraldpay/dshackle/upstream/rpcclient/stream/Responses.kt [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+data class SingleResponse(
+    val result: ByteArray?,
+    val error: JsonRpcError?,
+) : Response() {
+    fun hasError() = error != null
+
+    fun noResponse() = result == null && error == null
+
+    override fun equals(other: Any?): Boolean {
```
Comment (a10zn8) [inline src/main/kotlin/io/emeraldpay/dshackle/upstream/rpcclient/stream/Responses.kt:16]: Why we need that on data classes in this file?

File: src/main/kotlin/io/emeraldpay/dshackle/upstream/rpcclient/stream/JsonRpcStreamParser.kt [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.atomic.AtomicReference
+
+class JsonRpcStreamParser {
+    companion object {
+        private val jsonFactory = JsonFactory()
+        private val responseRpcParser = ResponseRpcParser()
+
+        private const val ARRAY_OPEN_BRACKET: Byte = 91
```
Comment (a10zn8) [inline src/main/kotlin/io/emeraldpay/dshackle/upstream/rpcclient/stream/JsonRpcStreamParser.kt:21]: Maybe use smth like `'['.code.toByte()` for better readability ?

File: src/main/kotlin/io/emeraldpay/dshackle/upstream/rpcclient/stream/JsonRpcStreamParser.kt [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+                .map { bytes ->
+                    val whatCountValue = whatCount.get()
+                    for (i in bytes.indices) {
+                        if (whatCountValue is Count.CountObjectBrackets) {
+                            countBrackets(bytes[i], count, OBJECT_OPEN_BRACKET, OBJECT_CLOSE_BRACKET)
+                        } else if (whatCountValue is Count.CountArrayBrackets) {
+                            countBrackets(bytes[i], count, ARRAY_OPEN_BRACKET, ARRAY_CLOSE_BRACKET)
+                        } else if (whatCountValue is Count.CountQuotes) {
+                            if (bytes[i] == QUOTE) {
+                                count.decrementAndGet()
```
Comment (a10zn8) [inline src/main/kotlin/io/emeraldpay/dshackle/upstream/rpcclient/stream/JsonRpcStreamParser.kt:116]: So, technically you can meet something like \" inside string value, so looks like this parsing could be broken

File: src/main/kotlin/io/emeraldpay/dshackle/upstream/rpcclient/stream/JsonRpcStreamParser.kt [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+    private fun processScalarValue(
+        token: JsonToken,
+        tokenStart: Int,
+        bytes: ByteArray,
+        endStream: AtomicBoolean,
+    ): ByteArray? {
+        if (token == JsonToken.VALUE_NULL) {
+            endStream.set(true)
+            return "null".toByteArray()
+        } else if (token == JsonToken.VALUE_STRING) {
```
Comment (a10zn8) [inline src/main/kotlin/io/emeraldpay/dshackle/upstream/rpcclient/stream/JsonRpcStreamParser.kt:226]: Looks like there are could be numeric or boolean value too.
Maybe add some log for default case?

**Review by a10zn8 — APPROVED** [review]

File: src/main/kotlin/io/emeraldpay/dshackle/upstream/calls/DefaultEthereumMethods.kt [hunk 5]
```diff
@@ -129,6 +129,8 @@ class DefaultEthereumMethods(
         "eth_chainId",
     )
 
+    private val streamMethods = setOf("eth_call", "eth_getLogs")
```
[the remark this answers is not in the document]
Comment (KirillPamPam) [reply src/main/kotlin/io/emeraldpay/dshackle/upstream/calls/DefaultEthereumMethods.kt]: Turned on for all methods. If something happen we can turn it off from our side on dproxy

File: src/main/kotlin/io/emeraldpay/dshackle/upstream/rpcclient/stream/JsonRpcStreamParser.kt (hunk 2 above)
Comment (KirillPamPam) [reply src/main/kotlin/io/emeraldpay/dshackle/upstream/rpcclient/stream/JsonRpcStreamParser.kt]: Cool idea, done, thanks!

File: src/main/kotlin/io/emeraldpay/dshackle/upstream/rpcclient/stream/JsonRpcStreamParser.kt (hunk 3 above)
Comment (KirillPamPam) [reply src/main/kotlin/io/emeraldpay/dshackle/upstream/rpcclient/stream/JsonRpcStreamParser.kt:118]: Fixed

File: src/main/kotlin/io/emeraldpay/dshackle/upstream/rpcclient/stream/JsonRpcStreamParser.kt (hunk 4 above)
Comment (KirillPamPam) [reply src/main/kotlin/io/emeraldpay/dshackle/upstream/rpcclient/stream/JsonRpcStreamParser.kt]: Added `else` branch for all others

**Review by a10zn8 — APPROVED** [review]

File: src/main/kotlin/io/emeraldpay/dshackle/upstream/rpcclient/stream/JsonRpcStreamParser.kt [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+                .filter { !endStream.get() }
+                .map { bytes ->
+                    val whatCountValue = whatCount.get()
+                    for (i in bytes.indices) {
+                        if (whatCountValue is Count.CountObjectBrackets) {
+                            countBrackets(bytes[i], count, OBJECT_OPEN_BRACKET, OBJECT_CLOSE_BRACKET)
+                        } else if (whatCountValue is Count.CountArrayBrackets) {
+                            countBrackets(bytes[i], count, ARRAY_OPEN_BRACKET, ARRAY_CLOSE_BRACKET)
+                        } else if (whatCountValue is Count.CountQuotes) {
+                            if (bytes[i] == QUOTE && bytes[i - 1] != BACKSLASH) {
```
Comment (Termina1) [inline src/main/kotlin/io/emeraldpay/dshackle/upstream/rpcclient/stream/JsonRpcStreamParser.kt:117]: what if bytes[i-2] also slash? You need to count slashes the same way as brackets.

```
if (bytes[I] == SLASH) {
  slashCounter = slashCounter ^ 1
}

and then when you find quote you just need to check that slashCounter == 0

**Review by Termina1 — APPROVED** [review]

**Pull request merged**